src/main/java/it/pagopa/pm/gateway/controller/PostePayPaymentTransactionsController.java

```diff
@@ -342,4 +362,100 @@ private String getCustomEnvironmentProperty(String parameterizedPropertyName, St
+                        + detailsPaymentRequest.getPaymentID() + " error message " + e.getMessage());
+            }
+        }
+
+        if (!isAuthCodeEmpty || checkDetail) {
+            return executeRefundRequest(detailsPaymentRequest, requestEntity);
+        }
+
+        return createPostePayRefundResponse(requestId, requestEntity.getCorrelationId(), null, ExceptionsEnum.TRANSACTION_REFUND_NOT_AUTHORIZED);
+
+    }
+
+    private ResponseEntity<PostePayRefundResponse> executeRefundRequest(DetailsPaymentRequest detailsPaymentRequest, PaymentRequestEntity requestEntity) {
+        InlineResponse2002 inlineResponse2002;
+        log.info("START - execute refund PostePay request for paymentRequestEntity with guid: " + requestEntity.getGuid());
+
+        try {
+            inlineResponse2002 = postePayControllerApi.apiV1PaymentRefundPost(detailsPaymentRequest);
+            if (Objects.isNull(inlineResponse2002)) {
+                log.error("Response for request refund for paymentId: " + detailsPaymentRequest.getPaymentID() + " is null");
+                return createPostePayRefundResponse(requestEntity.getGuid(), requestEntity.getCorrelationId(), null, ExceptionsEnum.PSP_CLIENT_EXCEPTION);
+            }
+            requestEntity.setIsRefund(inlineResponse2002.getTransactionResult().equals(EsitoStorno.OK));
+            paymentRequestRepository.save(requestEntity);
+            return createPostePayRefundResponse(requestEntity.getGuid(), requestEntity.getCorrelationId(), inlineResponse2002.getTransactionResult().getValue(), null);
+
+        } catch (ApiException e) {
+            log.error("Exception while calling Postepay PSP for refund DetailsPaymentRequest with paymentId: " + detailsPaymentRequest.getPaymentID());
+
+            if (e.getCode() == HttpStatus.REQUEST_TIMEOUT.value()) {
+                return createPostePayRefundResponse(requestEntity.getGuid(), requestEntity.getCorrelationId(), null, ExceptionsEnum.TIMEOUT);
+            }
+            return createPostePayRefundResponse(requestEntity.getGuid(), requestEntity.getCorrelationId(), null, ExceptionsEnum.PSP_CLIENT_EXCEPTION);
+        } catch (Exception e) {
+            log.error("Generic Exception while refund DetailsPaymentRequest with paymentId: " + detailsPaymentRequest.getPaymentID());
+            return createPostePayRefundResponse(requestEntity.getGuid(), requestEntity.getCorrelationId(), null, ExceptionsEnum.GENERIC_ERROR);
+
+        }
+
+    }
+
+    private boolean checkDetailStatus(DetailsPaymentRequest detailsPaymentRequest) throws Exception {
+        InlineResponse2001 inlineResponse2001;
+        inlineResponse2001 = postePayControllerApi.apiV1PaymentDetailsPost(detailsPaymentRequest);
+        if (Objects.isNull(inlineResponse2001)) {
+            throw new RestApiException(ExceptionsEnum.PSP_CLIENT_EXCEPTION);
+        }
+
+        Esito esito = inlineResponse2001.getStatus();
+        return !Objects.isNull(esito) && esito.equals(Esito.APPROVED);
```

Comment:
!Objects.isNull(esito) can be written as:
```suggestion
        return Objects.nonNull(esito)
```